stl/inc/numbers

```diff
@@ -0,0 +1,98 @@
+// numbers standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _NUMBERS_
+#define _NUMBERS_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+#if !_HAS_CXX20
+#pragma message("The contents of <numbers> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#ifdef __cpp_lib_concepts
+#include <concepts>
+#else // ^^^ defined(__cpp_lib_concepts) / !defined(__cpp_lib_concepts) vvv
+#include <type_traits>
+#endif
+
+_STD_BEGIN
+namespace numbers {
+#ifdef __cpp_lib_concepts
+    template <floating_point _Floating>
+    inline constexpr _Floating e_v = static_cast<_Floating>(2.71828182845904523536);
+    template <floating_point _Floating>
+    inline constexpr _Floating log2e_v = static_cast<_Floating>(1.44269504088896340736);
+    template <floating_point _Floating>
+    inline constexpr _Floating log10e_v = static_cast<_Floating>(0.434294481903251827651);
+    template <floating_point _Floating>
+    inline constexpr _Floating pi_v = static_cast<_Floating>(3.14159265358979323846);
+    template <floating_point _Floating>
+    inline constexpr _Floating inv_pi_v = static_cast<_Floating>(0.318309886183790671538);
+    template <floating_point _Floating>
+    inline constexpr _Floating inv_sqrtpi_v = static_cast<_Floating>(0.564189583547756286948);
+    template <floating_point _Floating>
+    inline constexpr _Floating ln2_v = static_cast<_Floating>(0.693147180559945309417);
+    template <floating_point _Floating>
+    inline constexpr _Floating ln10_v = static_cast<_Floating>(2.30258509299404568402);
+    template <floating_point _Floating>
+    inline constexpr _Floating sqrt2_v = static_cast<_Floating>(1.41421356237309504880);
+    template <floating_point _Floating>
+    inline constexpr _Floating sqrt3_v = static_cast<_Floating>(1.73205080756887729353);
+    template <floating_point _Floating>
+    inline constexpr _Floating inv_sqrt3_v = static_cast<_Floating>(0.577350269189625764509);
+    template <floating_point _Floating>
+    inline constexpr _Floating egamma_v = static_cast<_Floating>(0.577215664901532860607);
+    template <floating_point _Floating>
+    inline constexpr _Floating phi_v   = static_cast<_Floating>(1.61803398874989484820);
+#else // ^^^ defined(__cpp_lib_concepts) / !defined(__cpp_lib_concepts) vvv
+    template <class _Ty>
+    using _Floating = enable_if_t<is_floating_point_v<_Ty>, _Ty>;
+
+    template <class _Ty>
+    inline constexpr _Ty e_v = static_cast<_Floating<_Ty>>(2.71828182845904523536);

```

Comment:
Do we expect users to say `e_v<int>` or whatever, and get a confusing diagnostic about "can't initialize `int` from `void`"? I wonder whether a `static_assert` strategy would be better, like:

https://github.com/microsoft/STL/blob/d9cf06ead0f12479c2a6fce3135ae521d870df55/stl/inc/array#L196-L205

(This emits a `static_assert` in a deduction guide, which doesn't directly provide a way to `static_assert`, just like these variable templates.)

However, as this is C++20 non-concepts code, I would be fine with the code as-is.

Reply:
I did consider a `static_assert` though it's my understanding that one of the purported benefits of concepts is providing sane error messages.

So by providing a `static_assert` it would be providing on par or even better messages than the concepts version which currently just `the associated constraints are not satisfied. see declaration...`. I didn't think non-concepts users deserved that 😛 

(though on a serious note I figured it would be short-lived as you say so wouldn't matter much)